runtime/src/status_cache.rs

```diff
@@ -122,186 +162,361 @@ impl<T: Serialize + Clone> StatusCache<T> {
-        if let Some(stored_forks) = keymap.get(key_slice) {
-            let res = stored_forks
+
+        txs.get(key_slice).and_then(|forks| {
+            forks
                 .iter()
-                .find(|(f, _)| ancestors.contains_key(f) || self.roots.contains(f))
-                .cloned();
-            if res.is_some() {
-                return res;
-            }
-        }
-        None
+                .find(|(slot, _)| ancestors.contains_key(slot) || self.roots.contains(slot))
+                .cloned()
+        })
     }
 
-    /// Search for a key with any blockhash
-    /// Prefer get_status for performance reasons, it doesn't need
-    /// to search all blockhashes.
+    /// Search for a key with any blockhash.
+    ///
+    /// Prefer get_status for performance reasons, it doesn't need to search all blockhashes.
     pub fn get_status_any_blockhash<K: AsRef<[u8]>>(
         &self,
         key: K,
         ancestors: &Ancestors,
     ) -> Option<(Slot, T)> {
-        let keys: Vec<_> = self.cache.keys().copied().collect();
-
-        for blockhash in keys.iter() {
-            trace!("get_status_any_blockhash: trying {}", blockhash);
-            let status = self.get_status(&key, blockhash, ancestors);
+        for item in self.cache.iter() {
+            let (_, key_index, txs) = &**item.value();
+            let status = self.do_get_status(txs, *key_index, &key, ancestors);
             if status.is_some() {
                 return status;
             }
         }
         None
     }
 
-    /// Add a known root fork.  Roots are always valid ancestors.
-    /// After MAX_CACHE_ENTRIES, roots are removed, and any old keys are cleared.
-    pub fn add_root(&mut self, fork: Slot) {
+    /// Add a known root fork.
+    ///
+    /// Roots are always valid ancestors. After MAX_CACHE_ENTRIES, roots are removed, and any old
+    /// keys are cleared.
+    pub fn add_root(&self, fork: Slot) {
         self.roots.insert(fork);
         self.purge_roots();
     }
 
-    pub fn roots(&self) -> &HashSet<Slot> {
-        &self.roots
+    /// Get all the roots.
+    pub fn roots(&self) -> impl Iterator<Item = Slot> + '_ {
+        self.roots.iter().map(|x| *x)
     }
 
-    /// Insert a new key for a specific slot.
-    pub fn insert<K: AsRef<[u8]>>(
-        &mut self,
-        transaction_blockhash: &Hash,
-        key: K,
-        slot: Slot,
-        res: T,
-    ) {
+    /// Insert a new key using the given blockhash at the given slot.
+    pub fn insert<K: AsRef<[u8]>>(&self, blockhash: &Hash, key: K, slot: Slot, res: T) {
         let max_key_index = key.as_ref().len().saturating_sub(CACHED_KEY_SIZE + 1);
+        let mut key_slice = MaybeUninit::<[u8; CACHED_KEY_SIZE]>::uninit();
 
         // Get the cache entry for this blockhash.
-        let (max_slot, key_index, hash_map) =
-            self.cache.entry(*transaction_blockhash).or_insert_with(|| {
+        let key_index = {
```

Comment:
If we returned from this block `(key_index, key_slice)` would that perform a copy, or would compiler optimize it to effectively what you have now?

Thinking about how we can limit the scope of `MaybeUninit` so easier to maintain. Not opposed to what you have if there's **any** performance cost to my suggestion.

Basically change the inner spot where you're currently doing ptr copy to something like this:
```rust
            // Grab the key slice.
            let key_index = (*key_index).min(max_key_index);
            let key_slice = {
                let mut key_slice = MaybeUninit::<[u8; CACHED_KEY_SIZE]>::uninit();
                unsafe {
                    ptr::copy_nonoverlapping(
                        key.as_ref()[key_index..key_index + CACHED_KEY_SIZE].as_ptr(),
                        key_slice.as_mut_ptr() as *mut u8,
                        CACHED_KEY_SIZE,
                    );
                    key_slice.assume_init()
                }
            };

            // Insert the slot and tx result into the cache entry associated with
            // this blockhash and keyslice.
            let mut forks = txs.entry(key_slice).or_default();
            forks.push((slot, res.clone()));

            (key_index, key_slice)
```

now all the unsafe is in one place.
IF this introduces extra copies, let's just keep it as is. I think the compiler should be smart enough to do what we want, but probably putting too much faith in it 😢